lwip_demo/lwip_demo.c

```diff
@@ -0,0 +1,333 @@
+   IP6_ADDR(&ip6addr, 1234, 5666, 1914, 5111);
+   IP4_ADDR(&netmask, 255, 255, 255, 0);
+
+   netif_control = netif_add(&netif, &ipaddr, &netmask, &gw, NULL, my_init,
+                             netif_input);
+
+   if (netif_control == NULL) {
+      printf("netif_add_ip6_address failed.\n");
+   }
+
+   err_ipv6_netif = netif_add_ip6_address(&netif, &ip6addr, NULL);
+   if (err_ipv6_netif != ERR_OK) {
+      printf("netif_add_ip6_address failed.\n");
+   }
+
+   netif.name[0] = 'e';
+   netif.name[1] = '0';
+
+   netif_create_ip6_linklocal_address(&netif, 1);
+   netif.ip6_autoconfig_enabled = 1;
+
+   netif_set_status_callback(&netif, netif_status_callback);
+   netif_set_default(&netif);
+   netif_set_up(&netif);
+
+#if LWIP_DHCP
+   dhcp_init(&netif);
+#endif // LWIP_DHCP
+
+   // All initialization done, we're ready to receive data
+   printf("main: Reset done, Init done, Interrupts enabled\n");
+   printf("main: IER Register: %x\n", *IER);
+
+   optimsoc_list_iterator_t iter = 0;
+   eth_rx_pbuf_queue = optimsoc_list_init(NULL);
+   eth_rx_pbuf_queue = NULL;
+
+#if LWIP_UDP
+   static struct udp_pcb *udpecho_raw_pcb;
+   udp_init();
+   udpecho_raw_pcb = udp_new_ip_type(IPADDR_TYPE_ANY);
+   udp_recv(udpecho_raw_pcb, NULL, NULL);
+   udp_bind_netif(udpecho_raw_pcb, &netif);
+#endif // LWIP_UDP
+
+#if LWIP_DEBUG
+   debug_flags |= (LWIP_DBG_ON|LWIP_DBG_TRACE|LWIP_DBG_STATE|LWIP_DBG_FRESH|LWIP_DBG_HALT); // |LWIP_DBG_TRACE
+#endif //LWIP_DEBUG
+
+#if LWIP_TCP
+   static struct tcp_pcb *tcpecho_raw_pcb;
+   tcpecho_raw_pcb = tcp_new_ip_type(IPADDR_TYPE_ANY);
+   if (tcpecho_raw_pcb != NULL) {
+      tcp_bind(tcpecho_raw_pcb, IP_ANY_TYPE, 2049);
+      tcp_bind_netif(tcpecho_raw_pcb, &netif);
+   }
+#endif // LWIP_TCP
+
+   while (1) {
+      // TODO: Check link status
+
+      if (eth_rx_pbuf_queue != NULL
+            && optimsoc_list_length(eth_rx_pbuf_queue) != 0) {
+         if (NULL == optimsoc_list_first_element(eth_rx_pbuf_queue, &iter)) {
+            // printf("main: Element was NULL, return!\n");
+            eth_rx_pbuf_queue = NULL;
+         } else {
+            uint32_t restore = or1k_critical_begin();
+            struct pbuf* p = (struct pbuf*) optimsoc_list_remove_head(
+                  eth_rx_pbuf_queue);
+            or1k_critical_end(restore);
+
+            LINK_STATS_INC(link.recv);MIB2_STATS_NETIF_ADD(&netif, ifoutoctets, p->tot_len);
+            int unicast = ((((uint16_t *) p->payload)[0] & 0x01) == 0);
+            if (unicast) {
+               MIB2_STATS_NETIF_INC(&netif, ifoutucastpkts);
+            } else {
+               MIB2_STATS_NETIF_INC(&netif, ifoutnucastpkts);
+            }
+
+            // TODO: activate the checksum test
+
+            if (netif.input(p, &netif) != ERR_OK) {
+               pbuf_free(p);
+            } else {
+               eth_rx_pbuf_queue = NULL;
+            }
+         }
+
+      }
+
+      for (int i = 0; i <= 1000; i++)
+         ; // For loop (busy waiting)
```

Comment:
did you figure out why this is necessary? Please add a comment describing why we have this code (or if you don't know why it works, just describe what would happen otherwise)